csharp/ql/src/utils/model-generator/internal/CaptureTypeBasedSummaryModels.qll

```diff
@@ -0,0 +1,231 @@
+ * can be provided as input to `callable`.
+ * This includes
+ * (1) The implicit synthetic field(s) of the declaring type of `callable`.
+ * (2) The parameters of `callable`.
+ * (3) Any delegate parameters of `callable`.
+ */
+private predicate input(DotNet::Callable callable, string input, TypeParameter tp) {
+  input = implicit(callable, tp)
+  or
+  parameter(callable, input, tp)
+  or
+  source(callable, input, tp)
+}
+
+/**
+ * Holds if `callable` returns a value of type `tp` (or a generic parameterized over `tp`) and `output`
+ * is a models as data string representation of, how data is routed to the return.
+ */
+private predicate returns(DotNet::Callable callable, TypeParameter tp, string output) {
+  exists(Type return, string access | return = callable.getReturnType() |
+    access = getAccess(callable, return, tp) and
+    output = "ReturnValue" + access
+  )
+}
+
+/**
+ * Holds if `callable` has a delegate parameter that accepts a value of type `tp`
+ * and `output` is the models as data string representation of, how data is routed to
+ * the delegate parameter.
+ */
+private predicate sink(DotNet::Callable callable, TypeParameter tp, string output) {
+  exists(DelegateType dt, int position, Type t, Parameter p |
+    delegate(callable, dt, position) and
+    p = dt.getAParameter() and
+    t = p.getType() and
+    t = tp and
+    output = "Argument[" + position + "]" + ".Parameter[" + p.getPosition() + "]"
+  )
+}
+
+/**
+ * Holds if `output` is a models as data string representation of, how values of type `tp`
+ * (or generics parameterized over `tp`) can be routed.
+ * This includes
+ * (1) The implicit synthetic field(s) of the declaring type of `callable`.
+ * (2) The return of `callable`.
+ * (3) Any delegate parameters of `callable`.
+ */
+private predicate output(DotNet::Callable callable, TypeParameter tp, string output) {
+  output = implicit(callable, tp)
+  or
+  returns(callable, tp, output)
+  or
+  sink(callable, tp, output)
+}
+
+/**
+ * A class of callables that are relevant generating summaries for based
+ * on the Theorems for Free approach.
+ */
+class TypeBasedFlowTargetApi extends Specific::TargetApiSpecific {
+  TypeBasedFlowTargetApi() { Specific::isRelevantForTypeBasedFlowModels(this) }
+
+  /**
+   * Gets the string representation of all type based summaries inspired by
+   * the Theorems for Free approach.
+   *
+   * Basic example signatures could be
+   * this : T -> \alpha
+   * this : \beta -> T
+   * where T is type parameter on `this` or on the declaring type of `this`.
+   *
+   * Important special cases are \alpha = unit (setter),
+   * \alpha = T (getter, setter and id) and \beta = unit (getter).
+   *
+   * Complex example signatures could be
+   * this : (T -> S) -> S
+   * this : S1 x (S1 -> S2) -> S2
+   * where T is type parameter of the class declaring `this` and S, S1 and S2 are type parameters
+   * of `this`.
+   */
+  string getSummaries() {
```

Comment:
What if we have
```csharp
void M<T>(Func<T,T> f)
```
do we then get a flow summary that passes the output of calling `f` into `f` itself? I would think we shouldn't generate such a summary.

Reply:
Yes, we would get a summary for that. 
Do you dislike summaries for the example above because (1) there is no other source that could *generate* a value of type `T` or (2) we don't expect fix-point calculations 😂 ?
In the case you mention here it might not make sense, but we do need to generate summaries like this under some circumstances (think of `fold` or in terms of C# `Aggregate`).

Reply:
Ah, yes, it is indeed needed for `Aggregate`.